gno/p/role_manager/role_manager.gno

```diff
@@ -0,0 +1,187 @@
+		panic("caller does not have permission")
+	}
+
+	if !rm.roles.Has(roleName) {
+		panic("role does not exist")
+	}
+
+	role := rm.mustGetRole(roleName)
+	role.users.Iterate("", "", func(key string, value interface{}) bool {
+		user := std.Address(key)
+		userRoles := rm.getUser(user)
+		userRoles.Remove(roleName)
+		return false
+	})
+	rm.roles.Remove(roleName)
+}
+
+func (rm *RoleManager) AddPermissionToRole(permission string, roleName string) {
+	caller := std.PrevRealm().Addr()
+	if !rm.HasPermission(caller, PermissionWriteRole) && rm.owner.Owner() != caller {
+		panic("caller does not have permission")
+	}
+
+	role := rm.mustGetRole(roleName)
+	role.permissions.Set(permission, struct{}{})
+}
+
+func (rm *RoleManager) RemovePermissionFromRole(permission string, roleName string) {
+	caller := std.PrevRealm().Addr()
+	if !rm.HasPermission(caller, PermissionWriteRole) && rm.owner.Owner() != caller {
+		panic("caller does not have permission")
+	}
+
+	role := rm.mustGetRole(roleName)
+	role.permissions.Remove(permission)
+}
+
+func (rm *RoleManager) AddRoleToUser(user std.Address, roleName string) {
+	caller := std.PrevRealm().Addr()
+	if !rm.HasPermission(caller, PermissionWriteRole) && rm.owner.Owner() != caller {
+		panic("caller does not have permission")
+	}
+
+	role := rm.mustGetRole(roleName)
+	userRoles := rm.getUser(user)
+	userRoles.Set(roleName, role)
+	role.users.Set(user.String(), struct{}{})
+}
+
+func (rm *RoleManager) RemoveRoleFromUser(user std.Address, roleName string) {
+	caller := std.PrevRealm().Addr()
+	if !rm.HasPermission(caller, PermissionWriteRole) && rm.owner.Owner() != caller {
+		panic("caller does not have permission")
+	}
+
+	role := rm.mustGetRole(roleName)
+	userRoles := rm.getUser(user)
+	userRoles.Remove(roleName)
+	role.users.Remove(user.String())
+}
+
+func (rm *RoleManager) HasPermission(user std.Address, permission string) bool {
+	userRoles := rm.getUser(user)
+	res := false
+	userRoles.Iterate("", "", func(key string, value interface{}) bool {
+		role, ok := value.(*Role)
+		if !ok {
+			panic("role does not exist")
+		}
+		if role.permissions.Has(permission) || role.permissions.Has(PermissionAll) {
+			res = true
+			return true
+		}
+		return false
+	})
+	return res
+}
+
+func (rm *RoleManager) HasRole(user std.Address, roleName string) bool {
+	userRoles := rm.getUser(user)
+	return userRoles.Has(roleName)
+}
+
+func (rm *RoleManager) mustGetRole(roleName string) *Role {
+	role, ok := rm.roles.Get(roleName)
+	if !ok {
+		panic("role does not exist")
+	}
+	return role.(*Role)
+}
+
+// if user does not exist, create it with the user role
+func (rm *RoleManager) getUser(addr std.Address) *avl.Tree {
+	user, ok := rm.users.Get(addr.String())
+	if !ok {
+		uRole := rm.mustGetRole(RoleUser)
+		tree := avl.NewTree()
+		tree.Set(RoleUser, uRole)
+		rm.users.Set(addr.String(), tree)
+		uRole.users.Set(addr.String(), struct{}{})
+		return tree
+	}
+	return user.(*avl.Tree)
```

Comment:
mutating state in getters is misleading

Reply:
resolve [aa249ec](https://github.com/TERITORI/teritori-dapp/pull/1293/commits/aa249ec4a581a01be7407b0490bca3f87c25c564)

Reply:
you still add roles in HasPermission, HasRole which is a getter

I also find weird that you add roles in DeleteRole and stuff like that

I think we should `initUser` only in `AddRoleToUser`